habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py

```diff
@@ -505,33 +496,58 @@ def get_all_scenedataset_receptacles(
     return receptacles
 
 
-def import_tri_mesh_ply(ply_file: str) -> TriangleMeshData:
+def import_tri_mesh(mesh_file: str) -> List[mn.trade.MeshData]:
     """
-    Returns a Tuple of (verts,indices) from a ply mesh using magnum trade importer.
+    Returns a list of MeshData objects from a mesh asset using magnum trade importer.
 
-    :param ply_file: The input PLY mesh file. NOTE: must contain only triangles.
+    :param mesh_file: The input meshes file. NOTE: must contain only triangles.
     """
     manager = mn.trade.ImporterManager()
     importer = manager.load_and_instantiate("AnySceneImporter")
-    importer.open_file(ply_file)
+    importer.open_file(mesh_file)
 
-    # TODO: We don't support mesh merging or multi-mesh parsing currently
-    if importer.mesh_count > 1:
-        raise NotImplementedError(
-            "Importing multi-mesh receptacles (mesh merging or multi-mesh parsing) is not supported."
-        )
+    mesh_data: List[mn.trade.MeshData] = []
 
-    mesh_ix = 0
-    mesh = importer.mesh(mesh_ix)
-    assert (
-        mesh.primitive == mn.MeshPrimitive.TRIANGLES
-    ), "Must be a triangle mesh."
+    if importer.scene_count > 0:
+        scene_id = importer.default_scene
+        # If there's no default scene, load the first one
+        if scene_id == -1:
+            scene_id = 0
 
-    # zero-copy reference to importer datastructures
-    mesh_data = TriangleMeshData(
-        mesh.attribute(mn.trade.MeshAttribute.POSITION),
-        mesh.indices,
-    )
+        scene = importer.scene(scene_id)
+
+        # Mesh referenced by mesh_assignments[i] has a corresponding transform in
+        # mesh_transformations[i]. Association to a particular node ID is stored in
+        # scene.mapping(mn.trade.SceneField.MESH)[i], but it's not needed for anything
+        # here.
+        mesh_assignments: cr.containers.StridedArrayView1D = scene.field(
+            mn.trade.SceneField.MESH
+        )
+        mesh_transformations: List[
+            mn.Matrix4
+        ] = mn.scenetools.flatten_transformation_hierarchy3d(
+            scene, mn.trade.SceneField.MESH
+        )
+        assert len(mesh_assignments) == len(mesh_transformations)
+
+        # A mesh can be referenced by multiple nodes, so this can't operate in-place.
+        flattened_meshes: List[mn.trade.MeshData] = [
+            mn.meshtools.transform3d(importer.mesh(mesh_id), transformation)
+            for mesh_id, transformation in zip(
+                mesh_assignments, mesh_transformations
+            )
+        ]
+
+        for mesh in flattened_meshes:
+            assert (
+                mesh.primitive == mn.MeshPrimitive.TRIANGLES
+            ), "Must be a triangle mesh."
```

Comment:
There's also a new [meshtools.generate_indices()](https://doc.magnum.graphics/python/magnum/meshtools/#generate_indices) utility that can convert `TRIANGLE_STRIP` and `TRIANGLE_FAN` primitives to `TRIANGLES`. Those aren't that common, but some datasets have them, so it'd be an additional step for robustness:

```py
            if mesh.primitive in [mn.MeshPrimitive.TRIANGLE_STRIP, mn.MeshPrimitive.TRIANGLE_FAN]:
                mesh = mn.meshtools.generate_indices(mesh)
            assert (
                mesh.primitive == mn.MeshPrimitive.TRIANGLES
            ), "Must be a triangle mesh."
```

[See C++ docs for details.](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1MeshTools.html#a8b0d9114105a36fbd2147ee129e2017d)

Hm, also probably best to do this during the initial import at the very top, again to not have to do that several times for the same mesh in case of duplicated chairs and such.

Reply:
Thanks, applied this on the initial import step